src/peripherals/openlog.c

```
@@ -0,0 +1,524 @@
+
+    if (data_read) {
+      if (response_index < response_size) {
+        response[response_index++] = c;
+      }
+
+      if (terminators_match(c, terminators, terminator_indexes, terminators_count)) {
+        // null-terminate response (note response_size may be zero)
+        if (response_index+1 < response_size) {
+          response_index++;
+        }
+        if (response_index < response_size) {
+          response[response_index] = 0;
+        }
+        return true;
+      }
+    }
+
+    timeout = (chTimeNow() - start_ticks) >= MS2ST(wait_max_ms);
+    if (timeout) {
+      return false;
+    }
+
+    if (!data_read) {
+      chThdSleepMilliseconds(RECEIVE_DELAY_ms);
+    }
+  } while (!timeout);
+
+  return false;
+}
+
+static bool usart_wait_for_str(usart_dma_state *s, const char *str, u32 wait_max_ms)
+{
+  return usart_wait_for_response(s, 0, 0, &str, 1, wait_max_ms);
+}
+
+static bool usart_wait_for_strs(usart_dma_state *s, const char **strs, u32 strs_count, u32 wait_max_ms)
+{
+  return usart_wait_for_response(s, 0, 0, strs, strs_count, wait_max_ms);
+}
+
+static bool command_execute(usart_dma_state *s, bool (*func)(usart_dma_state *s, void *params), void *params)
+{
+  u32 retry = 0;
+  bool success = false;
+
+  do {
+      success = func(s, params);
+  } while (!success && (++retry < COMMAND_RETRY_COUNT));
+
+  return success;
+}
+
+static bool strs_concat(char *dest, u32 dest_len, const char **srcs, u32 srcs_count)
+{
+  u32 dest_index = 0;
+  u32 srcs_index;
+
+  for (srcs_index = 0; srcs_index < srcs_count; srcs_index++) {
+    const char *src = srcs[srcs_index];
+
+    u32 i = 0;
+    while (src[i] != 0) {
+      if (dest_index < dest_len) {
+        dest[dest_index++] = src[i++];
+      } else {
+        return false;
+      }
+    }
+  }
+
+  // null terminate
+  if (dest_index < dest_len) {
+    dest[dest_index++] = 0;
+    return true;
+  } else {
+    return false;
+  }
+}
+
+static bool parse_int(s32 *output, const char *str)
```

They silently return zero on failure.